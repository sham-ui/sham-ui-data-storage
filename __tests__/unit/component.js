import renderer, { compileAsSFC, compile } from 'sham-ui-test-helpers';
import createStorage from '../../src/index';

describe( 'component', () => {
    const updating = () => new Promise( resolve => setImmediate( resolve ) );

    it( 'base', async() => {
        const { storage, useStorage }  = createStorage( {
            firstName: 'John',
            lastName: 'Smith'
        } );

        const meta = renderer(
            compileAsSFC( {
                useStorage
            } )`
                <template>
                    <dl>
                        <dt>First name</dt>
                        <dd>{{user.firstName}}</dd>
                        <dt>Last name</dt>
                        <dd>{{user.lastName}}</dd>
                    </dl>
                </template>
                <script>
                    @useStorage( 'user' )
                    class Component extends Template {}
                    export default Component;
                </script>
            `
        );
        expect( meta.toJSON() ).toMatchSnapshot( 'initial' );
        storage.firstName = 'Jordan';
        expect( meta.toJSON() ).toMatchSnapshot( 'after update firstName' );
        await updating();
        expect( meta.toJSON() ).toMatchSnapshot( 'after update firstName & wait updating' );
        storage.lastName = 'Shah';
        expect( meta.toJSON() ).toMatchSnapshot( 'after update lastName' );
        expect( meta.toJSON() ).toMatchSnapshot( 'after update lastName & wait updating' );
        storage.lastName = 'Shah';
        await updating();
        expect( meta.toJSON() ).toMatchSnapshot( 'lastName don\'t changed' );
    } );

    it( 'remove', async() => {
        const { storage, useStorage }  = createStorage( {
            firstName: 'John',
            lastName: 'Smith'
        } );

        const UserInfoComponent = compileAsSFC( {
            useStorage
        } )`
            <template>
                <dl>
                    <dt>First name</dt>
                    <dd>{{user.firstName}}</dd>
                    <dt>Last name</dt>
                    <dd>{{user.lastName}}</dd>
                </dl>
            </template>
            <script>
                @useStorage( 'user' )
                class Component extends Template {}
                export default Component;
            </script>
        `;

        const meta = renderer(
            compile( {
                UserInfo: UserInfoComponent
            } )`
                {% if test %}
                    <UserInfo/>
                {% endif %}
            `,
            {
                test: true
            }
        );
        expect( meta.toJSON() ).toMatchSnapshot( 'initial' );
        storage.firstName = 'Jordan';
        await updating();
        expect( meta.toJSON() ).toMatchSnapshot( 'after update firstName' );
        meta.component.update( {
            test: false
        } );
        expect( meta.toJSON() ).toMatchSnapshot( 'test === false' );
        storage.lastName = 'Shah';
        await updating();
        expect( meta.toJSON() ).toMatchSnapshot( 'after update lastName' );
        meta.component.update( {
            test: true
        } );
        expect( meta.toJSON() ).toMatchSnapshot( 'test === true' );
        storage.firstName = 'John';
        storage.lastName = 'Smith';
        await updating();
        expect( meta.toJSON() ).toMatchSnapshot( 'after firstName & lastName' );
    } );

    it( 'sync', () => {
        const { storage, useStorage }  = createStorage( {
            firstName: 'John',
            lastName: 'Smith'
        } );

        const meta = renderer(
            compileAsSFC( {
                useStorage
            } )`
                <template>
                    <dl>
                        <dt>First name</dt>
                        <dd>{{user.firstName}}</dd>
                        <dt>Last name</dt>
                        <dd>{{user.lastName}}</dd>
                    </dl>
                </template>
                <script>
                    @useStorage( 'user' )
                    class Component extends Template {}
                    export default Component;
                </script>
            `
        );
        storage.firstName = 'Jordan';
        expect( meta.toJSON() ).toMatchSnapshot( 'after update firstName' );
        storage.sync();
        expect( meta.toJSON() ).toMatchSnapshot( 'after update firstName & wait updating' );
        storage.lastName = 'Shah';
        expect( meta.toJSON() ).toMatchSnapshot( 'after update lastName' );
        expect( meta.toJSON() ).toMatchSnapshot( 'after update lastName & wait updating' );
        storage.lastName = 'Shah';
        storage.sync();
        expect( meta.toJSON() ).toMatchSnapshot( 'lastName don\'t changed' );
    } );

} );
