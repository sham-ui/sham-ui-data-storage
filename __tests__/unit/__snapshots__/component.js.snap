// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component base: after update firstName & wait updating 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Smith",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component base: after update firstName 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Smith",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>John</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component base: after update lastName & wait updating 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component base: after update lastName 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component base: initial 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "John",
      "lastName": "Smith",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>John</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component base: lastName don't changed 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Shah</dd>
</dl>
",
}
`;

exports[`component remove: after firstName & lastName 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": true,
    "types": Array [],
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>John</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
<!--UserInfo-->
<!--if-->
",
}
`;

exports[`component remove: after update firstName 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": true,
    "types": Array [],
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
<!--UserInfo-->
<!--if-->
",
}
`;

exports[`component remove: after update lastName 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": false,
    "types": Array [],
  },
  "html": "<!--if-->",
}
`;

exports[`component remove: initial 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": true,
    "types": Array [],
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>John</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
<!--UserInfo-->
<!--if-->
",
}
`;

exports[`component remove: test === false 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": false,
    "types": Array [],
  },
  "html": "<!--if-->",
}
`;

exports[`component remove: test === true 1`] = `
Object {
  "Constructor": "dummy",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "test": true,
    "types": Array [],
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Shah</dd>
</dl>
<!--UserInfo-->
<!--if-->
",
}
`;

exports[`component sync: after update firstName & wait updating 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Smith",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component sync: after update firstName 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Smith",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>John</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component sync: after update lastName & wait updating 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component sync: after update lastName 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Smith</dd>
</dl>
",
}
`;

exports[`component sync: lastName don't changed 1`] = `
Object {
  "Constructor": "ComponentWithStorage",
  "Options": Object {
    "ID": "component",
    "containerSelector": "body",
    "types": Array [],
    "user": Object {
      "firstName": "Jordan",
      "lastName": "Shah",
    },
  },
  "html": "
<dl>
  <dt>First name</dt>
  <dd>Jordan</dd>
  <dt>Last name</dt>
  <dd>Shah</dd>
</dl>
",
}
`;
